types/styled-components/index.d.ts

```diff
@@ -141,6 +144,8 @@ export interface StyledComponentBase<
     //   props: StyledComponentPropsWithAs<AsC, T, O, A>
     // ): React.ReactElement<StyledComponentPropsWithAs<AsC, T, O, A>>
 
+    defaultProps: ReactDefaultProps<C>;
```

Comment:
I'd add `readonly` here

Reply:
done

Reply:
This is because of how the type copying works.

The `extends` clause will _only_ accept components that have already strictly defined `defaultProps`; if your component has a partial `defaultProps?: foo` it won't pick it up and will create a `defaultProps: never` key. You can't assign anything to it anyway.

If we accept partial `defaultProps?:`, then the `defaultProps` copying we are doing here is useless for `LibraryManagedAttributes` purposes, which is what this bug is intended to fix. Either we have to add a partial prop that will be ignored by `LibraryManagedAttributes`, or we have to lie and say that the `defaultProps` _are_ defined even if you don't initialize them, which is also wrong and the real issue I wanted to avoid by making it `readonly`.

This is because TypeScript doesn't support expando properties in TypeScript, only in JavaScript. There's no way in real TypeScript to have a property assignment or function call modify the type of its `this`, which is what we need to make `defaultProps` _actually_ work as you intended to.

Reply:
This is slightly unfortunate as setting the defaultProps of a styled component is a fairly common pattern, at least in my experience. 

This is also more of a breaking change without a workaround (unless you count `// @ts-ignore`) for existing usage of styled-components.

Reply:
I wasn't aware that TypeScript doesn't support expando properties, although when I think about it that is understandable.

The issue still stands that `defaultProps` is not `readonly` in reality and for me this is a breaking change. Do you think there's a way around this other than `ts-ignore`?

Reply:
Let me try to use this pattern and see if it helps. Any code snippet would be useful.

Reply:
The workaround would be `.attrs`, but that has its own bugs (https://github.com/styled-components/styled-components/issues/2246) :(

Reply:
I second the feeling here, just updated the typings and i set defaultProps all the time, so it's an issue. I'll stick a `@ts-ignore` on it for now.